drivers/interrupt_controller/intc_exti_stm32.c

```diff
@@ -0,0 +1,372 @@
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include "stm32_hsem.h"
+#include "intc_exti_stm32_priv.h"
+
+LOG_MODULE_REGISTER(exti_stm32, CONFIG_INTC_LOG_LEVEL);
+
+#define IS_VALID_EXTI_LINE_NUM(line_num) ((line_num) < STM32_EXTI_TOTAL_LINES_NUM)
+
+/*
+ * The boilerplate for COND_CODE_x is needed because the values are not 0/1
+ */
+#if STM32_EXTI_TOTAL_LINES_NUM > 32
+#define HAS_LINES_32_63	1
+#if STM32_EXTI_TOTAL_LINES_NUM > 64
+#define HAS_LINES_64_95	1
+#endif /* STM32_EXTI_TOTAL_LINES_NUM > 64 */
+#endif /* STM32_EXTI_TOTAL_LINES_NUM > 32 */
+
+#define EXTI_FN_HANDLER(_fn, line_num, line)			\
+	if (line_num < 32U) {								\
+		_fn(0_31, line);								\
+IF_ENABLED(HAS_LINES_32_63, (							\
+	} else if (line_num < 64U) {						\
+		_fn(32_63, line);								\
+))										\
+IF_ENABLED(HAS_LINES_64_95, (							\
+	} else if (line_num < 96U) {						\
+		_fn(64_95, line);								\
+))										\
+	} else {							\
+		LOG_ERR("Invalid line number %u", line_num);	\
+		__ASSERT_NO_MSG(0);								\
+	}
+
+
+bool stm32_exti_is_pending(uint32_t line_num)
+{
+	bool ret = false;
+	const uint32_t line = exti_linenum_to_ll_exti_line(line_num);
+
+	if (!IS_VALID_EXTI_LINE_NUM(line_num)) {
+		LOG_ERR("Invalid line number %u", line_num);
+		return false;
+	}
+
+	z_stm32_hsem_lock(CFG_HW_EXTI_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	/**
+	 * Note: we can't use EXTI_FN_HANDLER here because we care
+	 * about the return value of EXTI_IS_ACTIVE_FLAG.
+	 */
+	if (line_num < 32U) {
+		ret = EXTI_IS_ACTIVE_FLAG(0_31, line);
+#if STM32_EXTI_TOTAL_LINES_NUM > 32
+	} else if (line_num < 64U) {
+		ret = EXTI_IS_ACTIVE_FLAG(32_63, line);
+#endif /* STM32_EXTI_TOTAL_LINES_NUM > 32 */
+#if STM32_EXTI_TOTAL_LINES_NUM > 64
+	} else if (line_num < 96U) {
+		ret = EXTI_IS_ACTIVE_FLAG(64_95, line);
+#endif /* STM32_EXTI_TOTAL_LINES_NUM > 64 */
+	} else {
+		LOG_ERR("Invalid line number: %u", line_num);
+		__ASSERT_NO_MSG(0);
+	}
+
+	z_stm32_hsem_unlock(CFG_HW_EXTI_SEMID);
+
+	return ret;
+}
+
+int stm32_exti_clear_pending(uint32_t line_num)
+{
+	const uint32_t line = exti_linenum_to_ll_exti_line(line_num);
+
+	if (!IS_VALID_EXTI_LINE_NUM(line_num)) {
+		LOG_ERR("Invalid line number %u", line_num);
+		return -EINVAL;
+	}
+
+	z_stm32_hsem_lock(CFG_HW_EXTI_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	EXTI_FN_HANDLER(EXTI_CLEAR_FLAG, line_num, line);
+
+	z_stm32_hsem_unlock(CFG_HW_EXTI_SEMID);
+
+	return 0;
+}
+
+int stm32_exti_sw_interrupt(uint32_t line_num)
+{
+	const uint32_t line = exti_linenum_to_ll_exti_line(line_num);
+
+	if (!IS_VALID_EXTI_LINE_NUM(line_num)) {
+		LOG_ERR("Invalid line number %u", line_num);
+		return -EINVAL;
+	}
+
+	z_stm32_hsem_lock(CFG_HW_EXTI_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+
+	EXTI_FN_HANDLER(EXTI_GENERATE_SWI, line_num, line);
+
+	z_stm32_hsem_unlock(CFG_HW_EXTI_SEMID);
+
+	return 0;
+}
+
+/** Enables the peripheral clock required to access EXTI registers */
+static int stm32_exti_enable_registers(void)
```

Comment:
```suggestion
static int stm32_exti_enable_clocks(void)
```

Reply:
Can't apply suggestion any more cause I've rebased over the new master first. But I've applied this change manually. Please take a look.